[v1.14] Proxy ports persistence fixes 1.14

jrajahalme opened the pull request:
* [x] #35839 (@jrajahalme)

Once this PR is merged, a GitHub action will update the labels of these PRs:
```upstream-prs
 35839
```

jrajahalme committed:
proxy: Ensure ports are stored on shutdown

[ upstream commit 1dd7b63e54b67c4f6ae985d80c936878eb8723a2 ]

This commit wraps the proxy ports checkpoint function in a controller. This
has two effects:

  1. If the checkpoint fails due to I/O errors, the checkpoint is
     retried until it succeeds. Previously, we relied on it being
     triggered again for it to be re-ran (though so far we have not
     observed any I/O failures in CI, so this likely was not an issue)
  2. By using a controller `StopFunc` and calling `RemoveControllerAndWait`
     we ensure the checkpoint function is ran one last time during
     shutdown. This is important, since `Trigger.Shutdown` by itself does not
     run any pending triggers. Note that any errors during shutdown are
     only logged and the controller is not re-tried.

The proxy ports trigger is closed when the proxy cell is
stopped, which happens e.g. when SIGTERM is received.

[ Backporter's notes: import conflicts manually resolved in
  `pkg/proxy/cell.go` file. ]

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
Signed-off-by: viktor-kurchenko <viktor.kurchenko@isovalent.com>

tommyp1ckles commented on daemon/cmd/daemon_main.go:
```diff
@@ -440,6 +440,9 @@ func initializeFlags() {
 	option.BindEnv(Vp, option.ProxyIdleTimeout)
 
+	flags.Uint(option.RestoredProxyPortsAgeLimit, 15, "Time after which a restored proxy ports file is considered stale (in minutes)")
+	option.BindEnv(vp, option.RestoredProxyPortsAgeLimit)
```
Some compile failures, think we need to manually change this `vp` -> `Vp`

jrajahalme replied:
Was not the first time I've been bitten by this... fixed!

jrajahalme committed:
proxy: Restore proxy ports in cell, skip stale proxy ports file

[ upstream commit 70ec9b39c6d99070732ab97d3148fd07e84c45da ]

Move proxy port restoration to the cell just before the trigger for
writing the proxy ports file is created. This removes the possibility
that the trigger would run before the ports are restored.

Add a configurable time limit ('--restored-proxy-ports-age-limit',
default 15 minutes) for the age of the restored proxy ports file to
remove the chance that stale port numbers would be used e.g., on upgrade
after a downgrade to Cilium 1.15 (which does not store the proxy ports to
a file).

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
Signed-off-by: viktor-kurchenko <viktor.kurchenko@isovalent.com>
